public/app/features/dashboard-scene/serialization/layoutSerializers/utils.ts

```diff
@@ -193,32 +193,49 @@ export function createPanelDataProvider(panelKind: PanelKind): SceneDataProvider
   });
 }
 
+/**
+ * Get panel-level datasource for a v2beta1 panel.
+ *
+ * In v2beta1 schema, there's NO panel-level datasource concept - each query has its own.
+ * However, we still need to set panel-level datasource to "mixed" when queries use
+ * different datasources, so the Scene can properly handle mixed datasource mode.
+ *
+ * This function returns:
+ * - Mixed datasource if queries use different datasources
+ * - undefined otherwise (each query has its own datasource)
+ *
+ * This ensures v2→Scene→v1 conversion produces the same output as the Go backend,
+ * which does NOT add panel-level datasource for non-mixed panels.
+ */
 function getPanelDataSource(panel: PanelKind): DataSourceRef | undefined {
   if (!panel.spec.data?.spec.queries?.length) {
     return undefined;
   }
 
-  let datasource: DataSourceRef | undefined = undefined;
+  let firstDatasource: DataSourceRef | undefined = undefined;
   let isMixedDatasource = false;
 
   panel.spec.data.spec.queries.forEach((query) => {
-    if (!datasource) {
-      if (!query.spec.query.datasource?.name) {
-        datasource = getRuntimePanelDataSource(query.spec.query);
-      } else {
-        datasource = {
-          uid: query.spec.query.datasource?.name,
-          type: query.spec.query.group,
-        };
-      }
-    } else if (datasource.uid !== query.spec.query.datasource?.name || datasource.type !== query.spec.query.group) {
+    const queryDs = query.spec.query.datasource?.name
+      ? { uid: query.spec.query.datasource.name, type: query.spec.query.group }
+      : getRuntimePanelDataSource(query.spec.query);
+
+    if (!firstDatasource) {
+      firstDatasource = queryDs;
+    } else if (firstDatasource.uid !== queryDs?.uid || firstDatasource.type !== queryDs?.type) {
       isMixedDatasource = true;
     }
   });
 
-  return isMixedDatasource ? { type: 'mixed', uid: MIXED_DATASOURCE_NAME } : datasource;
+  // Only return mixed datasource - for non-mixed panels, each query already has its own datasource
+  // This matches the Go backend behavior which doesn't add panel.datasource for non-mixed panels
+  return isMixedDatasource ? { type: 'mixed', uid: MIXED_DATASOURCE_NAME } : undefined;
```

Comment:
Why is this code changed? It's responsible for finding a data source for SQR, has really nothing to do with serialization. I'm slighlty worried that changes like this are source for regressions:

https://github.com/grafana/grafana/issues/114636
https://github.com/grafana/grafana/issues/114638

Reply:
I believe this change caused this issue - https://github.com/grafana/grafana/issues/115269